src/Proper/HasPermutationGenerator.hs

```diff
@@ -0,0 +1,301 @@
+                           [(fromString $ dup <> " not unique", property $ failure)
+                           | dup <- dups]
+                          ]
+           else [Group "HasPermutationGenerator Graph Not Strongly Connected" $
+                          [(fromString "Not strongly connected", abortNotSCC ns graph)]
+                        ]
+    where
+      abortNotSCC ns graph =
+        let (a,b) = findNoPath (Proxy :: Proxy m) ns graph
+          in property $ failWithFootnote $ renderStyle ourStyle $
+               "PermutationEdges do not form a strongly connected graph."
+               $+$ hang "No Edge Between here:" 4 (ppDoc a)
+               $+$ hang "            and here:" 4 (ppDoc b)
+      findDupEdgeNames = [ name g | g <- generators :: [PermutationEdge m p]
+                                  , length (filter (==g) generators) > 1 ]
+      testEdge :: Map Int (Set p)
+               -> Map (Int,Int) [PermutationEdge m p]
+               -> (Set p -> PropertyT IO m)
+               -> PermutationEdge m p
+               -> Group
+      testEdge ns pem mGen pe =
+        Group (fromString (name pe)) $ (fromString "Is Required", runRequiredTest):
+          [ (edgeTestName f t, runEdgeTest f t)
+          | (f,t) <- matchesEdges
+          ]
+        where
+          matchesEdges = [ e | (e,v) <- Map.toList pem, pe `elem` v ]
+          edgeTestName f t = fromString $ name pe <> " : " <> (show $ Set.toList (lut ns f)) <> " -> " <> (show $ Set.toList (lut ns t))
+          isRequired =
+            let x = [ () | (_,v) <- Map.toList pem, length v == 1, p <- v, p == pe ]
+             in length x > 0
+          runRequiredTest = property $ do
+            if isRequired
+               then pure ()
+               else failWithFootnote $ renderStyle ourStyle $
+                      (fromString $ "PermutationEdge " <> name pe <> " is not required to make graph strongly connected.")
+                      $+$ hang "Edge:" 4 (ppDoc $ name pe)
+          runEdgeTest f t = property $ do
+            om <- mGen (lut ns f)
+            nm <- forAll $ (permuteGen pe) om
+            let expected = lut ns t
+                observed = properties nm
+            if expected == observed
+              then pure ()
+              else failWithFootnote $ renderStyle ourStyle $
+                     "PermutationEdge fails its contract."
+                       $+$ hang "Edge:" 4 (ppDoc $ name pe)
+                       $+$ hang "Expected:" 4 (ppDoc expected)
+                       $+$ hang "Observed:" 4 (ppDoc observed)
+
+  buildGen :: forall t . Monad t => Gen m -> Set p -> PropertyT t m
+  buildGen g = do
+    let pedges = findPermutationEdges (Proxy :: Proxy m) (Proxy :: Proxy p)
+        edges = Map.keys pedges
+        distmap = distanceMap edges
+        (sn,ns) = numberNodes (Proxy :: Proxy m) (Proxy :: Proxy p)
+        graph = buildGraph pedges
+        isco = isStronglyConnected graph
```

Comment:
We compute the graph and distance matrix up front when we build the generator which allows us to traverse the graph over random shortest paths extremely quickly.